File: 449-serialize-and-deserialize-bst/449-serialize-and-deserialize-bst.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Codec {

    // Encodes a tree to a single string.
    public String serialize(TreeNode root) {
        if(root == null) return "";
        StringBuilder sb = new StringBuilder();
        dfs1(root, sb);
        return sb.toString();
    }
    void dfs1(TreeNode root, StringBuilder sb) {
        if(root == null) {
            return;
        }
        sb.append(root.val + ",");
        dfs1(root.left, sb);
        dfs1(root.right, sb);
    }
    

    // Decodes your encoded data to tree.
    public TreeNode deserialize(String data) {
        if(data == null || data.length() == 0) return null;
        Queue<String> q = new LinkedList<>(Arrays.asList(data.split(",")));
        return dfs2(q, Integer.MIN_VALUE, Integer.MAX_VALUE);
    }
    TreeNode dfs2(Queue<String> q, int min, int max) {
        if(q.isEmpty()) return null;
        String s = q.peek();
        int val = Integer.valueOf(s);
        if(val < min || val > max) return null;
        TreeNode node = new TreeNode(val);
        q.poll();
        node.left = dfs2(q, min, val);
        node.right = dfs2(q, val, max);
        return node;
    }
}

// Your Codec object will be instantiated and called as such:
// Codec ser = new Codec();
// Codec deser = new Codec();
// String tree = ser.serialize(root);
// TreeNode ans = deser.deserialize(tree);
// return ans;

File: 449-serialize-and-deserialize-bst/449-serialize-and-deserialize-bst.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Codec:

    def serialize(self, root: Optional[TreeNode]) -> str:
        """Encodes a tree to a single string.
        """
        if not root: return ''
        res = []
        def dfs(root):
            nonlocal res
            if not root:
                return 
            res.append(str(root.val))
            dfs(root.left)
            dfs(root.right)
        dfs(root)
        return ','.join(res)
    
    def deserialize(self, data: str) -> Optional[TreeNode]:
        """Decodes your encoded data to tree.
        """
        if not data: return None
        lst = deque(data.split(','))
        def dfs2(lst, min, max):
            val = int(lst[0])
            if val >= max or val <= min:
                return None
            root = TreeNode(lst.popleft())
            
            if lst: root.left = dfs2(lst, min, val)
            if lst: root.right = dfs2(lst, val, max)
            return root
        return dfs2(lst, -1, 10001)
# Your Codec object will be instantiated and called as such:
# Your Codec object will be instantiated and called as such:
# ser = Codec()
# deser = Codec()
# tree = ser.serialize(root)
# ans = deser.deserialize(tree)
# return ans